src/codegen/templates/_grpc_interpreter.py.mako

```diff
@@ -325,16 +411,105 @@ class GrpcStubInterpreter(BaseInterpreter):
+            _T0_EPOCH = ht_datetime(1904, 1, 1, tzinfo=timezone.utc)
+            # fractional_seconds is non-negative fractions of a second at 2^-64 resolution.
+            fractional_seconds_as_time = grpc_waveform.t0.fractional_seconds / 2**64
+            t0_seconds = grpc_waveform.t0.seconds + fractional_seconds_as_time
+            timestamp = _T0_EPOCH + ht_timedelta(seconds=t0_seconds)
+
+            if grpc_waveform.dt:
+                sample_interval = ht_timedelta(seconds=grpc_waveform.dt)
+            else:
+                sample_interval = ht_timedelta(seconds=0.0)
+
+            target_waveform.timing = Timing(
+                sample_interval_mode=SampleIntervalMode.REGULAR,
+                timestamp=timestamp,
+                sample_interval=sample_interval,
+            )
+
+def _copy_waveform_attributes(grpc_waveform, target_waveform):
+    if grpc_waveform.attributes:
+        for key, attr_value in grpc_waveform.attributes.items():
+            if attr_value.HasField('bool_value'):
+                target_waveform.extended_properties[key] = attr_value.bool_value
+            elif attr_value.HasField('integer_value'):
+                target_waveform.extended_properties[key] = attr_value.integer_value
+            elif attr_value.HasField('double_value'):
+                target_waveform.extended_properties[key] = attr_value.double_value
+            elif attr_value.HasField('string_value'):
+                target_waveform.extended_properties[key] = attr_value.string_value
+
+def _copy_protobuf_waveform_to_analog_waveform(grpc_waveform, target_waveform, waveform_attribute_mode):
+    samples_received = len(grpc_waveform.y_data)
+    _setup_waveform_capacity_and_samples(target_waveform, samples_received)
+    _assign_numpy_array(target_waveform.raw_data, grpc_waveform.y_data)
+    _copy_timing_information(grpc_waveform, target_waveform, waveform_attribute_mode)
+    _copy_waveform_attributes(grpc_waveform, target_waveform)
+
+def _copy_protobuf_waveform_to_digital_waveform(grpc_waveform, target_waveform, waveform_attribute_mode):
+    signal_count = grpc_waveform.signal_count
+    samples_received = len(grpc_waveform.y_data) // signal_count if signal_count > 0 else 0
+    _setup_waveform_capacity_and_samples(target_waveform, samples_received)
+    if samples_received > 0 and signal_count > 0:
+        data_array = numpy.frombuffer(grpc_waveform.y_data, dtype=numpy.uint8)
+        data_array = data_array.reshape(samples_received, signal_count)
+        target_waveform.data[:samples_received, :signal_count] = data_array
+    _copy_timing_information(grpc_waveform, target_waveform, waveform_attribute_mode)
+    _copy_waveform_attributes(grpc_waveform, target_waveform)
+
+def _copy_analog_waveform_to_protobuf_waveform(waveform, grpc_waveform):
+    scaled_data = waveform.scaled_data
+    if not scaled_data.flags.c_contiguous:
+        scaled_data = scaled_data.copy(order="C")
```

Comment:
Don't reinvent the wheel.

https://github.com/ni/ni-apis-python/blob/main/packages/ni.protobuf.types/src/ni/protobuf/types/waveform_conversion.py